Fails to compile on OTP 23 due to erl_interface.h

OTP 23 deprecated erl_interface and this package now fails with:

```erlang
===> Compiling /Users/codeadict/dev/bcrypt/c_src/bcrypt_port.c
===> /Users/codeadict/dev/bcrypt/c_src/bcrypt_port.c:20:10: fatal error: 'erl_interface.h' file not found
#include <erl_interface.h>
         ^~~~~~~~~~~~~~~~~
1 error generated.
```

--- a/.gitignore
+++ b/.gitignore
@@ -6,6 +6,8 @@ priv/bcrypt
 priv/bcrypt_nif.so
 logs/*
 .eunit
+rebar3
+.rebar
 
 ## libtool
 .deps

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,7 +1,10 @@
 language: erlang
 sudo: false
 otp_release:
-  - 20.0
+  - 23.0
+  - 22.3
+  - 21.3
+  - 20.3
   - 19.3
 
 jobs:

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,13 +4,11 @@ otp_release:
   - 23.0
   - 22.3
   - 21.3
-  - 20.3
-  - 19.3
 
 jobs:
   include:
     - stage: deploy
-      otp_release: 19.3
+      otp_release: 22.3
       script: skip
       deploy:
         provider: script

--- a/Makefile
+++ b/Makefile
@@ -1,12 +1,21 @@
-all: compile
-
-compile:
-	@rebar compile
-
-tests:
-	@rebar eunit
-
-clean:
-	@rebar clean
+REBAR := $(shell which rebar3 2>/dev/null || echo ./rebar3)
+REBAR_URL := https://s3.amazonaws.com/rebar3/rebar3
 
 .PHONY: clean compile tests
+
+all: compile
+
+compile: $(REBAR)
+	$(REBAR) compile
+
+tests: $(REBAR)
+	$(REBAR) eunit
+
+clean: $(REBAR)
+	$(REBAR) clean
+
+./rebar3:
+	erl -noshell -s inets start -s ssl start \
+		-eval '{ok, saved_to_file} = httpc:request(get, {"$(REBAR_URL)", []}, [], [{stream, "./rebar3"}])' \
+		-s inets stop -s init stop
+	chmod +x ./rebar3

--- a/c_src/Makefile
+++ b/c_src/Makefile
@@ -23,17 +23,16 @@ UNAME_SYS := $(shell uname -s)
 ifeq ($(UNAME_SYS), Darwin)
 	CC ?= cc
 	CFLAGS ?= -O3 -std=c99 -arch x86_64 -Wall -Wmissing-prototypes
-	CXXFLAGS ?= -O3 -arch x86_64 -Wall
 	LDFLAGS ?= -arch x86_64 -flat_namespace -undefined suppress
-	DRV_LDFLAGS = -flat_namespace -undefined suppress $(ERL_LDFLAGS) -lpthread
+	DRV_LDFLAGS = -flat_namespace -undefined suppress $(ERL_LDFLAGS)
 else ifeq ($(UNAME_SYS), FreeBSD)
 	CC ?= cc
-	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes -lpthread
-	CXXFLAGS ?= -O3 -Wall
+	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes
+	DRV_LDFLAGS = $(ERL_LDFLAGS)
 else ifeq ($(UNAME_SYS), Linux)
 	CC ?= gcc
-	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes -lpthread
-	CXXFLAGS ?= -O3 -Wall
+	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes
+	DRV_LDFLAGS = $(ERL_LDFLAGS)
 endif
 
 #   {"DRV_LDFLAGS","-shared $ERL_LDFLAGS -lpthread"},
@@ -44,10 +43,8 @@ endif
 #   {"LDFLAGS", "$LDFLAGS -lpthread"}
 
 CFLAGS += -fPIC -I $(ERTS_INCLUDE_DIR) -I $(ERL_INTERFACE_INCLUDE_DIR)
-CXXFLAGS += -fPIC -I $(ERTS_INCLUDE_DIR) -I $(ERL_INTERFACE_INCLUDE_DIR)
 
-LD_FLAGS += -lpthread
-LDLIBS += -L $(ERL_INTERFACE_LIB_DIR) -lei
+LDLIBS += -L $(ERL_INTERFACE_LIB_DIR) -lei -lpthread
 
 # Verbosity.
 

--- a/c_src/Makefile
+++ b/c_src/Makefile
@@ -1,9 +1,92 @@
-compile: ../priv/bcrypt
+# Based on c_src.mk from erlang.mk by Loic Hoguin <essen@ninenines.eu>
 
-../priv/bcrypt:
-	$(CC) $(CFLAGS) $(ERL_CFLAGS) $(EXE_LDFLAGS) bcrypt_port.c bcrypt.o blowfish.o $(LDFLAGS) $(ERL_LDFLAGS) -lpthread -o ../priv/bcrypt
+CURDIR := $(shell pwd)
+BASEDIR := $(abspath $(CURDIR)/..)
+
+PROJECT ?= $(notdir $(BASEDIR))
+PROJECT := $(strip $(PROJECT))
+
+ERTS_INCLUDE_DIR ?= $(shell erl -noshell -s init stop -eval "io:format(\"~s/erts-~s/include/\", [code:root_dir(), erlang:system_info(version)]).")
+ERL_INTERFACE_INCLUDE_DIR ?= $(shell erl -noshell -s init stop -eval "io:format(\"~s\", [code:lib_dir(erl_interface, include)]).")
+ERL_INTERFACE_LIB_DIR ?= $(shell erl -noshell -s init stop -eval "io:format(\"~s\", [code:lib_dir(erl_interface, lib)]).")
+
+C_SRC_DIR = $(CURDIR)
+C_SRC_NIF = $(CURDIR)/../priv/bcrypt_nif.so
+C_SRC_PORT = $(CURDIR)/../priv/bcrypt
+
+# DRV_LDFLAGS = -shared $(ERL_LDFLAGS) -lpthread
+DRV_CFLAGS = -Ic_src -Wall -O3 -fPIC $(ERL_CFLAGS)
+
+# System type and C compiler/flags.
+
+UNAME_SYS := $(shell uname -s)
+ifeq ($(UNAME_SYS), Darwin)
+	CC ?= cc
+	CFLAGS ?= -O3 -std=c99 -arch x86_64 -Wall -Wmissing-prototypes
+	CXXFLAGS ?= -O3 -arch x86_64 -Wall
+	LDFLAGS ?= -arch x86_64 -flat_namespace -undefined suppress
+	DRV_LDFLAGS = -flat_namespace -undefined suppress $(ERL_LDFLAGS) -lpthread
+else ifeq ($(UNAME_SYS), FreeBSD)
+	CC ?= cc
+	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes
+	CXXFLAGS ?= -O3 -Wall
+else ifeq ($(UNAME_SYS), Linux)
+	CC ?= gcc
+	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes
+	CXXFLAGS ?= -O3 -Wall
+endif
+
+#   {"DRV_LDFLAGS","-shared $ERL_LDFLAGS -lpthread"},
+#   {"darwin", "DRV_LDFLAGS", "-bundle -flat_namespace -undefined suppress $ERL_LDFLAGS -lpthread"},
+#   {"solaris", "ERL_LDFLAGS", "-lxnet -lssp -lnsl $ERL_LDFLAGS"},
+#   {"DRV_CFLAGS","-Ic_src -Wall -O3 -fPIC $ERL_CFLAGS"},
+#   {"CFLAGS", "$CFLAGS -Ic_src -Wall -O3"},
+#   {"LDFLAGS", "$LDFLAGS -lpthread"}
+
+CFLAGS += -fPIC -I $(ERTS_INCLUDE_DIR) -I $(ERL_INTERFACE_INCLUDE_DIR)
+CXXFLAGS += -fPIC -I $(ERTS_INCLUDE_DIR) -I $(ERL_INTERFACE_INCLUDE_DIR)
+
+LD_FLAGS += -lpthread
+LDLIBS += -L $(ERL_INTERFACE_LIB_DIR) -lei
+
+# Verbosity.
+
+c_verbose_0 = @echo " C     " $(?F);
+c_verbose = $(c_verbose_$(V))
+
+cpp_verbose_0 = @echo " CPP   " $(?F);
+cpp_verbose = $(cpp_verbose_$(V))
+
+link_verbose_0 = @echo " LD    " $(@F);
+link_verbose = $(link_verbose_$(V))
+
+SOURCES := $(shell find $(C_SRC_DIR) -type f \( -name "*.c" -o -name "*.C" -o -name "*.cc" -o -name "*.cpp" \))
+OBJECTS = $(addsuffix .o, $(basename $(SOURCES)))
+
+COMPILE_C = $(c_verbose) $(CC) $(CFLAGS) $(CPPFLAGS) -c
+COMPILE_CPP = $(cpp_verbose) $(CXX) $(CXXFLAGS) $(CPPFLAGS) -c
+
+all: $(C_SRC_NIF) $(C_SRC_PORT)
+
+$(C_SRC_PORT): $(OBJECTS)
+	@mkdir -p $(BASEDIR)/priv/
+	$(CC) $(CFLAGS) bcrypt_port.o bcrypt.o blowfish.o $(DRV_LDFLAGS) $(LDLIBS) -o ../priv/bcrypt
+
+$(C_SRC_NIF): $(OBJECTS)
+	@mkdir -p $(BASEDIR)/priv/
+	$(link_verbose) $(CC) $(OBJECTS) $(LDFLAGS) -shared $(LDLIBS) -o $(C_SRC_NIF)
+
+%.o: %.c
+	$(COMPILE_C) $(OUTPUT_OPTION) $<
+
+%.o: %.cc
+	$(COMPILE_CPP) $(OUTPUT_OPTION) $<
+
+%.o: %.C
+	$(COMPILE_CPP) $(OUTPUT_OPTION) $<
+
+%.o: %.cpp
+	$(COMPILE_CPP) $(OUTPUT_OPTION) $<
 
 clean:
-	@rm -f ../priv/bcrypt
-
-.PHONY: clean compile
+	@rm -f $(C_SRC_OUTPUT) $(OBJECTS)

--- a/c_src/Makefile
+++ b/c_src/Makefile
@@ -28,11 +28,11 @@ ifeq ($(UNAME_SYS), Darwin)
 	DRV_LDFLAGS = -flat_namespace -undefined suppress $(ERL_LDFLAGS) -lpthread
 else ifeq ($(UNAME_SYS), FreeBSD)
 	CC ?= cc
-	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes
+	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes -lpthread
 	CXXFLAGS ?= -O3 -Wall
 else ifeq ($(UNAME_SYS), Linux)
 	CC ?= gcc
-	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes
+	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes -lpthread
 	CXXFLAGS ?= -O3 -Wall
 endif
 

--- a/c_src/bcrypt_nif.c
+++ b/c_src/bcrypt_nif.c
@@ -24,14 +24,14 @@
 #include "erl_blf.h"
 #include "bcrypt_nif.h"
 
-void free_task(task_t* task)
+static void free_task(task_t* task)
 {
     if (task->env != NULL)
         enif_free_env(task->env);
     enif_free(task);
 }
 
-task_t* alloc_task(task_type_t type)
+static task_t* alloc_task(task_type_t type)
 {
     task_t* task = (task_t*)enif_alloc(sizeof(task_t));
     if (task == NULL)
@@ -41,7 +41,7 @@ task_t* alloc_task(task_type_t type)
     return task;
 }
 
-task_t* alloc_init_task(task_type_t type, ERL_NIF_TERM ref, ErlNifPid pid, int num_orig_terms, const ERL_NIF_TERM orig_terms[])
+static task_t* alloc_init_task(task_type_t type, ERL_NIF_TERM ref, ErlNifPid pid, int num_orig_terms, const ERL_NIF_TERM orig_terms[])
 {
     task_t* task = alloc_task(type);
     task->pid = pid;
@@ -102,7 +102,7 @@ static ERL_NIF_TERM hashpw(task_t* task)
         enif_make_string(task->env, encrypted, ERL_NIF_LATIN1));
 }
 
-void* async_worker(void* arg)
+static void* async_worker(void* arg)
 {
     ctx_t* ctx;
     task_t* task;

--- a/c_src/bcrypt_port.c
+++ b/c_src/bcrypt_port.c
@@ -17,39 +17,42 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <erl_interface.h>
-#include <ei.h>
 #include <unistd.h>
 
 #include "erl_blf.h"
+#include "ei.h"
 
 #define dec_int16(s) ((((unsigned char*)(s))[0] << 8) | \
                       (((unsigned char*)(s))[1]))
-
-#define enc_int16(i, s) {((unsigned char*)(s))[0] = ((i) >> 8) & 0xff;  \
-        ((unsigned char*)(s))[1] = (i) & 0xff;}
 
 #define BUFSIZE (1 << 16)
 #define CMD_SALT 0
 #define CMD_HASHPW 1
 
-typedef unsigned char byte;
+#define DATASIZE 1024
 
-int ts_bcrypt(char *, const char *, const char *);
-void encode_salt(char *, u_int8_t *, u_int16_t, u_int8_t);
+extern int ts_bcrypt(char *, const char *, const char *);
+extern void encode_salt(char *, u_int8_t *, u_int16_t, u_int8_t);
+
+
+static void
+fail(int place) {
+    fprintf(stderr, "Something went wrong %d\n", place);
+    exit(1);
+}
 
 /* These methods came from the Erlang port command tutorial:
  * http://www.erlang.org/doc/tutorial/c_port.html#4.2
  */
 static int
-read_buf(int fd, byte *buf, int len)
+read_buf(int fd, char *buf, int len)
 {
     int i, got = 0;
     do {
         if ((i = read(fd, buf+got, len-got)) <= 0) {
             if (i == 0) return got;
             if (errno != EINTR)
-                return got;
+                fail(-1);
             i = 0;
         }
         got += i;
@@ -58,7 +61,7 @@ read_buf(int fd, byte *buf, int len)
 }
 
 static int
-read_cmd(byte *buf)
+read_cmd(char *buf)
 {
     int len;
     if (read_buf(0, buf, 2) != 2)
@@ -69,165 +72,195 @@ read_cmd(byte *buf)
     return 1;
 }
 
-static int
-write_buf(int fd, byte *buf, int len)
+static void
+write_buf(int fd, const char *buf, int len)
 {
-    int i, done = 0;
+    int i;
+    int done = 0;
+
     do {
         if ((i = write(fd, buf+done, len-done)) < 0) {
             if (errno != EINTR)
-                return (i);
+                fail(-2);
             i = 0;
         }
         done += i;
     } while (done < len);
-    return (len);
 }
 
-static int
-write_cmd(byte *buf, int len)
+static void
+write_cmd(const char *buf, int len)
 {
-    byte hd[2];
-    enc_int16(len, hd);
-    if (write_buf(1, hd, 2) != 2)
-        return 0;
-    if (write_buf(1, buf, len) != len)
-        return 0;
-    return 1;
+    unsigned char li;
+
+    li = (len >> 8) & 0xff;
+    write_buf(1, (char *) &li, 1);
+    li = len & 0xff;
+    write_buf(1, (char *) &li, 1);
+    write_buf(1, buf, len);
 }
 
-static int
-process_reply(ETERM *pid, int cmd, const char *res)
+static void
+process_reply(int cmd, const char *result)
 {
-    ETERM *result;
-    int len, retval;
-    byte *buf;
-    result = erl_format("{~i, ~w, ~s}", cmd, pid, res);
-    len = erl_term_len(result);
-    buf = erl_malloc(len);
-    erl_encode(result, buf);
-    retval = write_cmd(buf, len);
-    erl_free_term(result);
-    erl_free(buf);
-    return retval;
+    ei_x_buff res_buf;
+
+    if (ei_x_new_with_version(&res_buf) != 0)
+        fail(-10);
+    if (ei_x_encode_tuple_header(&res_buf, 2) != 0)
+        fail(-11);
+    if (ei_x_encode_long(&res_buf, (long) cmd) != 0)
+        fail(-12);
+    if (ei_x_encode_binary(&res_buf, result, (long) strlen( (const char *) result)) != 0)
+        fail(-13);
+
+    write_cmd(res_buf.buff, res_buf.index);
+
+    if (ei_x_free(&res_buf) != 0)
+        fail(-14);
 }
 
-static int
-process_encode_salt(ETERM *pid, ETERM *data)
+static void
+process_error_reply(int cmd, const char *error)
 {
-    int retval = 0;
-    ETERM *pattern, *cslt, *lr;
-    byte *csalt = NULL;
-    long log_rounds = -1;
-    int csaltlen = -1;
-    char ret[64];
-    pattern = erl_format("{Csalt, LogRounds}");
-    if (erl_match(pattern, data)) {
-        cslt = erl_var_content(pattern, "Csalt");
-        csaltlen = ERL_BIN_SIZE(cslt);
-        csalt = ERL_BIN_PTR(cslt);
-        lr = erl_var_content(pattern, "LogRounds");
-        log_rounds = ERL_INT_UVALUE(lr);
-        if (16 != csaltlen) {
-            retval = process_reply(pid, CMD_SALT, "Invalid salt length");
-        } else if (log_rounds < 4 || log_rounds > 31) {
-            retval = process_reply(pid, CMD_SALT, "Invalid number of rounds");
-        } else {
-            encode_salt(ret, (u_int8_t*)csalt, csaltlen, log_rounds);
-            retval = process_reply(pid, CMD_SALT, ret);
-        }
-        erl_free_term(cslt);
-        erl_free_term(lr);
-    };
-    erl_free_term(pattern);
-    return retval;
+    ei_x_buff res_buf;
+
+    if (ei_x_new_with_version(&res_buf) != 0)
+        fail(-20);
+    if (ei_x_encode_tuple_header(&res_buf, 2) != 0)
+        fail(-21);
+    if (ei_x_encode_long(&res_buf, (long) cmd) != 0)
+        fail(-22);
+    if (ei_x_encode_atom(&res_buf, error) != 0)
+        fail(-23);
+
+    write_cmd(res_buf.buff, res_buf.index);
+
+    if (ei_x_free(&res_buf) != 0)
+        fail(-24);
 }
 
-static int
-process_hashpw(ETERM *pid, ETERM *data)
+static void
+process_encode_salt(
+    int salt_size,
+    char *salt,
+    long rounds)
 {
-    int retval = 0;
-    ETERM *pattern, *pwd, *slt, *pwd_bin, *slt_bin;
-    char password[1024];
-    char salt[1024];
-    char encrypted[1024] = { 0 };
+    char encoded_salt[64];
 
-    (void)memset(&password, '\0', sizeof(password));
-    (void)memset(&salt, '\0', sizeof(salt));
-
-    pattern = erl_format("{Pass, Salt}");
-    if (erl_match(pattern, data)) {
-        pwd = erl_var_content(pattern, "Pass");
-        pwd_bin = erl_iolist_to_binary(pwd);
-        slt = erl_var_content(pattern, "Salt");
-        slt_bin = erl_iolist_to_binary(slt);
-        if (ERL_BIN_SIZE(pwd_bin) > sizeof(password)) {
-            retval = process_reply(pid, CMD_HASHPW, "Password too long");
-        } else if (ERL_BIN_SIZE(slt_bin) > sizeof(salt)) {
-            retval = process_reply(pid, CMD_HASHPW, "Salt too long");
-        } else {
-            memcpy(password, ERL_BIN_PTR(pwd_bin), ERL_BIN_SIZE(pwd_bin));
-            memcpy(salt, ERL_BIN_PTR(slt_bin), ERL_BIN_SIZE(slt_bin));
-            if (ts_bcrypt(encrypted, password, salt)) {
-                retval = process_reply(pid, CMD_HASHPW, "Invalid salt");
-            } else {
-                retval = process_reply(pid, CMD_HASHPW, encrypted);
-            }
-        }
-        erl_free_term(pwd);
-        erl_free_term(slt);
-        erl_free_term(pwd_bin);
-        erl_free_term(slt_bin);
-    };
-    erl_free_term(pattern);
-    return retval;
-}
-
-static int
-process_command(unsigned char *buf)
-{
-    int retval = 0;
-    ETERM *pattern, *tuple, *cmd, *port, *data;
-    pattern = erl_format("{Cmd, Port, Data}");
-    tuple = erl_decode(buf);
-    if (erl_match(pattern, tuple)) {
-        cmd = erl_var_content(pattern, "Cmd");
-        port = erl_var_content(pattern, "Port");
-        data = erl_var_content(pattern, "Data");
-        switch (ERL_INT_VALUE(cmd)) {
-        case CMD_SALT:
-            retval = process_encode_salt(port, data);
-            break;
-        case CMD_HASHPW:
-            retval = process_hashpw(port, data);
-            break;
-        };
-        erl_free_term(cmd);
-        erl_free_term(port);
-        erl_free_term(data);
+    if (16 != salt_size) {
+        process_error_reply(CMD_SALT, "invalid_salt_length");
+    } else if (rounds < 4 || rounds > 31) {
+        process_error_reply(CMD_SALT, "invalid_rounds");
+    } else {
+        memset(encoded_salt, 0, 64);
+        encode_salt(encoded_salt, (u_int8_t *) salt, (u_int16_t) salt_size, (u_int8_t) rounds);
+        process_reply(CMD_SALT, encoded_salt);
     }
-    erl_free_term(pattern);
-    erl_free_term(tuple);
-    return retval;
+}
+
+static void
+process_hashpw(
+    int password_size,
+    char *password,
+    int salt_size,
+    char *salt)
+{
+    char encrypted[DATASIZE+1];
+
+    memset(encrypted, 0, DATASIZE+1);
+    if (ts_bcrypt(encrypted, password, salt)) {
+        process_error_reply(CMD_HASHPW, "invalid_salt");
+    } else {
+        process_reply(CMD_HASHPW, encrypted);
+    }
+}
+
+static void
+process_command(char *buf)
+{
+    int index = 0;
+    int version = 0;
+    int arity = 0;
+    int type;
+    long cmd;
+    long len;
+    long rounds;
+    int data_size;
+    char data[DATASIZE+1];
+    int salt_size;
+    char salt[DATASIZE+1];
+
+    memset(data, 0, DATASIZE+1);
+    memset(salt, 0, DATASIZE+1);
+
+    if (ei_decode_version(buf, &index, &version) != 0)
+        fail(1);
+
+    // Three tuple: {Cmd, Port, Data}
+    if (ei_decode_tuple_header(buf, &index, &arity) != 0)
+        fail(2);
+    if (arity != 2)
+        fail(3);
+    if (ei_decode_long(buf, &index, &cmd) != 0)
+        fail(4);
+
+    // All commands have a two tuple for Data
+    if (ei_decode_tuple_header(buf, &index, &arity) != 0)
+        fail(6);
+    if (arity != 2)
+        fail(7);
+
+    // First arg is always a binary
+    if (ei_get_type(buf, &index, &type, &data_size) != 0)
+        fail(8);
+    if (type != ERL_BINARY_EXT)
+        fail(9);
+    if (data_size < 0 || data_size > DATASIZE)
+        fail(10);
+    if (ei_decode_binary(buf, &index, data, &len) != 0)
+        fail(11);
+
+    switch (cmd) {
+        case CMD_HASHPW:
+            // Two tuple: {Pass, Salt}
+            if (ei_get_type(buf, &index, &type, &salt_size) != 0)
+                fail(12);
+            if (type != ERL_BINARY_EXT)
+                fail(13);
+            if (salt_size < 0 || salt_size > DATASIZE)
+                fail(14);
+            if (ei_decode_binary(buf, &index, salt, &len) != 0)
+                fail(15);
+
+            process_hashpw(data_size, data, salt_size, salt);
+            break;
+        case CMD_SALT:
+            // Two tuple: {Csalt, LogRounds}
+            if (ei_decode_long(buf, &index, &rounds) != 0)
+                fail(16);
+
+            process_encode_salt(data_size, data, rounds);
+            break;
+        default:
+            fail(17);
+    }
 }
 
 static void
 loop(void)
 {
-    byte buf[BUFSIZE];
-    int retval = 0;
-    do {
-        if (read_cmd(buf) > 0)
-            retval = process_command(buf);
-        else
-            retval = 0;
-    } while (retval);
+    char buf[BUFSIZE];
+
+    while (read_cmd(buf) == 1) {
+        process_command(buf);
+    }
 }
 
 int
 main(int argc, char *argv[])
 {
-    erl_init(NULL, 0);
+    ei_init();
     loop();
     return 0;
 }

--- a/rebar.config
+++ b/rebar.config
@@ -1,39 +1,10 @@
 %% -*- mode: erlang;erlang-indent-level: 2;indent-tabs-mode: nil -*-
 %% {erl_opts, [debug_info]}.
 
-%% {so_specs,
-%%  [{"priv/bcrypt_nif.so",
-%%    ["c_src/*.c"]}]}.
-{port_env, [
-  {"DRV_LDFLAGS","-shared $ERL_LDFLAGS -lpthread"},
-  {"darwin", "DRV_LDFLAGS", "-bundle -flat_namespace -undefined suppress $ERL_LDFLAGS -lpthread"},
-  {"solaris", "ERL_LDFLAGS", "-lxnet -lssp -lnsl $ERL_LDFLAGS"},
-  {"DRV_CFLAGS","-Ic_src -Wall -fPIC $ERL_CFLAGS"},
-  {"CFLAGS", "$CFLAGS -O2"},
-  {"LDFLAGS", "$LDFLAGS -lpthread"}
-]}.
+{pre_hooks,
+  [{"(linux|darwin|solaris)", compile, "make -C c_src"},
+   {"(freebsd)", compile, "gmake -C c_src"}]}.
 
-{port_specs, [
-    {"priv/bcrypt_nif.so", ["c_src/*.c"]},
-    {"priv/bcrypt", [
-        "c_src/bcrypt_port.c",
-        "c_src/bcrypt.c",
-        "c_src/blowfish.c"
-    ]}
-]}.
-
-%% These post_hooks are for rebar2. rebar.config.script removes them if
-%% rebar3 is detected
 {post_hooks,
- [{clean, "make -C c_src clean"},
-  {compile, "make -C c_src"}]}.
-
-%% plugins and provider_hooks are for rebar3. rebar.config.script removes them
-%% if rebar2 is detected
-{plugins, [pc]}.
-{provider_hooks, [
-  {pre, [
-    {compile, {pc, compile}},
-    {clean, {pc, clean}}
-  ]}
-]}.
+  [{"(linux|darwin|solaris)", clean, "make -C c_src clean"},
+   {"(freebsd)", clean, "gmake -C c_src clean"}]}.

--- a/rebar.config
+++ b/rebar.config
@@ -1,5 +1,9 @@
 %% -*- mode: erlang;erlang-indent-level: 2;indent-tabs-mode: nil -*-
-%% {erl_opts, [debug_info]}.
+
+{require_min_otp_vsn, "21"}.
+
+{erl_opts,
+  [debug_info]}.
 
 {pre_hooks,
   [{"(linux|darwin|solaris)", compile, "make -C c_src"},

--- a/rebar.config.script
+++ b/rebar.config.script
@@ -1,7 +1,0 @@
-case code:is_loaded(rebar3) of
-  false ->
-    C1 = proplists:delete(plugins, CONFIG),
-    proplists:delete(provider_hooks, C1);
-  {file, _} ->
-    proplists:delete(post_hooks, CONFIG)
-end.

--- a/src/bcrypt.erl
+++ b/src/bcrypt.erl
@@ -8,16 +8,31 @@
 -export([mechanism/0]).
 -export([gen_salt/0, gen_salt/1, hashpw/2]).
 
+-type mechanism() :: nif | port.
+-type rounds() :: 4..31.
+-type pwerr() :: invalid_salt | invalid_salt_length | invalid_rounds.
+
+-export_type([ mechanism/0, rounds/0, pwerr/0 ]).
+
 start() -> application:start(bcrypt).
 stop()  -> application:stop(bcrypt).
 
+-spec mechanism() -> mechanism().
 mechanism() ->
     {ok, M} = application:get_env(bcrypt, mechanism),
     M.
 
-gen_salt() -> do_gen_salt(mechanism()).
-gen_salt(Rounds) -> do_gen_salt(mechanism(), Rounds).
-hashpw(Password, Salt) -> do_hashpw(mechanism(), Password, Salt).
+-spec gen_salt() -> {ok, string()}.
+gen_salt() ->
+    do_gen_salt(mechanism()).
+
+-spec gen_salt( rounds() ) -> {ok, string()}.
+gen_salt(Rounds) ->
+    do_gen_salt(mechanism(), Rounds).
+
+-spec hashpw( string() | binary(), string() | binary() ) -> {ok, string()} | {error, pwerr()}.
+hashpw(Password, Salt) ->
+    do_hashpw(mechanism(), Password, Salt).
 
 do_gen_salt(nif)  -> bcrypt_nif_worker:gen_salt();
 do_gen_salt(port) -> bcrypt_pool:gen_salt().

--- a/src/bcrypt_nif_worker.erl
+++ b/src/bcrypt_nif_worker.erl
@@ -18,11 +18,15 @@
           context
          }).
 
-start_link() -> gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+start_link() ->
+  gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
 
-gen_salt() -> gen_server:call(?MODULE, gen_salt, infinity).
+gen_salt() ->
+  gen_server:call(?MODULE, gen_salt, infinity).
+
 gen_salt(Rounds) ->
     gen_server:call(?MODULE, {gen_salt, Rounds}, infinity).
+
 hashpw(Password, Salt) ->
     gen_server:call(?MODULE, {hashpw, Password, Salt}, infinity).
 
@@ -34,12 +38,14 @@ init([]) ->
 terminate(shutdown, _) -> ok.
 
 handle_call(gen_salt, _From, #state{default_log_rounds = R} = State) ->
-    {reply, {ok, bcrypt_nif:gen_salt(R)}, State};
+    Salt = bcrypt_nif:gen_salt(R),
+    {reply, {ok, Salt}, State};
 handle_call({gen_salt, R}, _From, State) ->
-    {reply, {ok, bcrypt_nif:gen_salt(R)}, State};
+    Salt = bcrypt_nif:gen_salt(R),
+    {reply, {ok, Salt}, State};
 handle_call({hashpw, Password, Salt}, _From, #state{context=Ctx}=State) ->
     Ref = make_ref(),
-    ok = bcrypt_nif:hashpw(Ctx, Ref, self(), Password, Salt),
+    ok = bcrypt_nif:hashpw(Ctx, Ref, self(), to_list(Password), to_list(Salt)),
     receive
         {ok, Ref, Result} ->
             {reply, {ok, Result}, State};
@@ -50,3 +56,6 @@ handle_call(Msg, _, _) -> exit({unknown_call, Msg}).
 handle_cast(Msg, _) -> exit({unknown_cast, Msg}).
 handle_info(Msg, _) -> exit({unknown_info, Msg}).
 code_change(_OldVsn, State, _Extra) -> {ok, State}.
+
+to_list(L) when is_list(L) -> L;
+to_list(B) when is_binary(B) -> binary_to_list(B).

--- a/src/bcrypt_port.erl
+++ b/src/bcrypt_port.erl
@@ -14,7 +14,11 @@
 -export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2,
          code_change/3]).
 
--record(state, {port, default_log_rounds}).
+-record(state, {
+    port :: port(),
+    default_log_rounds :: non_neg_integer(),
+    cmd_from :: {pid(), term()}
+    }).
 
 -define(CMD_SALT, 0).
 -define(CMD_HASH, 1).
@@ -71,38 +75,50 @@ terminate(_Reason, #state{port=Port}) ->
 
 handle_call({encode_salt, R}, From, #state{default_log_rounds = LogRounds} = State) ->
     handle_call({encode_salt, R, LogRounds}, From, State);
-handle_call({encode_salt, R, LogRounds}, From, State) ->
+handle_call({encode_salt, R, LogRounds}, From, #state{ cmd_from = undefined } = State) ->
     Port = State#state.port,
-    Data = term_to_binary({?CMD_SALT, From, {R, LogRounds}}),
-    port_command(Port, Data),
-    {noreply, State};
-handle_call({hashpw, Password, Salt}, From, State) ->
+    Data = term_to_binary({?CMD_SALT, {iolist_to_binary(R), LogRounds}}),
+    erlang:port_command(Port, Data),
+    {noreply, State#state{ cmd_from = From }};
+handle_call({encode_salt, _R, _Rounds}, From, #state{ cmd_from = CmdFrom } = State) ->
+    ?BCRYPT_ERROR("bcrypt: Salt request from ~p whilst busy for ~p", [ From, CmdFrom ]),
+    {reply, {error, {busy, From}}, State};
+handle_call({hashpw, Password, Salt}, From, #state{ cmd_from = undefined } = State) ->
     Port = State#state.port,
-    Data = term_to_binary({?CMD_HASH, From, {Password, Salt}}),
-    port_command(Port, Data),
-    {noreply, State};
+    Data = term_to_binary({?CMD_HASH, {iolist_to_binary(Password), iolist_to_binary(Salt)}}),
+    erlang:port_command(Port, Data),
+    {noreply, State#state{ cmd_from = From }};
+handle_call({hashpw, _Password, _Salt}, From, #state{ cmd_from = CmdFrom } = State) ->
+    ?BCRYPT_ERROR("bcrypt: Hash request from ~p whilst busy for ~p", [ From, CmdFrom ]),
+    {reply, {error, {busy, From}}, State};
 handle_call(stop, _From, State) ->
     {stop, normal, ok, State};
-handle_call(Msg, _, _) -> exit({unknown_call, Msg}).
+handle_call(Msg, _, State) ->
+    {stop, {unknown_call, Msg}, State}.
 
-handle_cast(Msg, _) -> exit({unknown_cast, Msg}).
+handle_cast(Msg, State) ->
+    {stop, {unknown_cast, Msg}, State}.
 
-code_change(_OldVsn, State, _Extra) -> {ok, State}.
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
 
-handle_info({Port, {data, Data}}, #state{port=Port}=State) ->
-    {Cmd, To, Reply0} = binary_to_term(Data),
+handle_info({Port, {data, Data}}, #state{ port = Port, cmd_from = From } = State) ->
     Reply =
-        case {Cmd, Reply0} of
-            {?CMD_SALT, "Invalid salt"} -> {error, invalid_salt};
-            {?CMD_SALT, "Invalid number of rounds"} -> {error, invalid_rounds};
-            {?CMD_HASH, "Invalid salt length"} -> {error, invalid_salt_length};
-            {_, _} when Cmd =:= ?CMD_SALT; Cmd =:= ?CMD_HASH -> {ok, Reply0}
+        case binary_to_term(Data) of
+            {_, Error} when is_atom(Error) ->
+                {error, Error};
+            {?CMD_SALT, Result} when is_binary(Result) ->
+                {ok, binary_to_list(Result)};
+            {?CMD_HASH, Result} when is_binary(Result) ->
+                {ok, binary_to_list(Result)}
         end,
-    gen_server:reply(To, Reply),
+    gen_server:reply(From, Reply),
     ok = bcrypt_pool:available(self()),
-    {noreply, State};
+    {noreply, State#state{ cmd_from = undefined }};
 handle_info({Port, {exit_status, Status}}, #state{port=Port}=State) ->
     %% Rely on whomever is supervising this process to restart.
     ?BCRYPT_WARNING("Port died: ~p", [Status]),
     {stop, port_died, State};
-handle_info(Msg, _) -> exit({unknown_info, Msg}).
+handle_info(Msg, _) ->
+    exit({unknown_info, Msg}).
+

--- a/test/bcrypt_tests.erl
+++ b/test/bcrypt_tests.erl
@@ -108,8 +108,12 @@ simple_nif_test_() ->
 
 pair_nif_test_() ->
     {setup, fun() -> ok = start_with(nif) end,
-     [?_assert({ok, Hash} =:= bcrypt:hashpw(Pass, Salt)) ||
-         {Pass, Salt, Hash} <- ?PAIRS]}.
+     [
+        begin
+            ?_assert({ok, Hash} =:= bcrypt:hashpw(Pass, Salt))
+        end
+        || {Pass, Salt, Hash} <- ?PAIRS
+     ]}.
 
 simple_port_test_() ->
     {setup, fun() -> ok = start_with(port) end,
@@ -123,5 +127,9 @@ simple_port_test_() ->
 
 pair_port_test_() ->
     {setup, fun() -> ok = start_with(port) end,
-     [?_assert({ok, Hash} =:= bcrypt:hashpw(Pass, Salt)) ||
-         {Pass, Salt, Hash} <- ?PAIRS]}.
+     [
+        begin
+            ?_assert({ok, Hash} =:= bcrypt:hashpw(Pass, Salt))
+        end
+        || {Pass, Salt, Hash} <- ?PAIRS
+     ]}.

